singularity-eos/closure/mixed_cell_models.hpp

```diff
@@ -647,6 +651,491 @@ class PTESolverRhoT : public mix_impl::PTESolverBase<EOSIndexer, RealIndexer> {
+}
+inline size_t PTESolverFixedTRequiredScratchInBytes(const int nmat) {
+  return PTESolverFixedTRequiredScratch(nmat) * sizeof(Real);
+}
+
+template <typename EOSIndexer, typename RealIndexer, typename LambdaIndexer>
+class PTESolverFixedT : public mix_impl::PTESolverBase<EOSIndexer, RealIndexer> {
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::AssignIncrement;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::nmat;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::neq;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::ResidualNorm;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::vfrac_total;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::eos;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::rho;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::vfrac;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::sie;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::temp;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::press;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::rho_total;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::u_total;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::MatIndex;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::jacobian;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::residual;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::dx;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::u;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::rhobar;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::Cache;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::Tnorm;
+
+ public:
+  // template the ctor to get type deduction/universal references prior to c++17
+  template <typename EOS_t, typename Real_t, typename Lambda_t>
+  PORTABLE_INLINE_FUNCTION
+  PTESolverFixedT(const int nmat, EOS_t &&eos, const Real vfrac_tot, const Real T_true,
+                Real_t &&rho, Real_t &&vfrac, Real_t &&sie, Real_t &&temp, Real_t &&press,
+                Lambda_t &&lambda, Real *scratch)
+      : mix_impl::PTESolverBase<EOSIndexer, RealIndexer>(nmat, nmat, eos, vfrac_tot,
+                                                         1.0, rho, vfrac, sie, temp,
+                                                         press, scratch) {
+    dpdv = AssignIncrement(scratch, nmat);
+    vtemp = AssignIncrement(scratch, nmat);
+    Tequil = T_true;
+    Ttemp = T_true;
+    Tnorm = 1.0;
+    // TODO(JCD): use whatever lambdas are passed in
+    /*for (int m = 0; m < nmat; m++) {
+      if (lambda[m] != nullptr) Cache[m] = lambda[m];
+    }*/
+  }
+
+  PORTABLE_INLINE_FUNCTION
+  Real Init() {
+    //InitBase();
+    // fake init base
+    // rhobar is a fixed quantity: the average density of
+    // material m averaged over the full PTE volume
+    rho_total = 0.0;
+    for (int m = 0; m < nmat; ++m) {
+      rhobar[m] = rho[m] * vfrac[m];
```

Comment:
Looks like you're sitll using `vfrac` here. Is that definitely what we want to pass in to this function?

Reply:
That's fine. They're `RealIndexer`s but. We can document which one we want. If we want to support both, we could use tag dispatch or something to specify which we're doing.

Reply:
I thought one thing we discussed was just overloading on the number of arguments so that you can specify `rhobar` in addition to `rho` and `vfrac`.

Another thing to consider is that `rhobar` and/or a mass fraction will be `const` whereas `rho` and `vfrac` are not. Could you overload on the `const`ness of these variables?

Reply:
> Figuring out constness of the underlying data of an indexer would not be trivial.

What I have done in the past is templated on both `ConstRealIndexer` and a `RealIndexer`. This doesn't enforce constness automatically, but if a const type is passed in, type inferrence notices that and treats it accordingly.

> I thought one thing we discussed was just overloading on the number of arguments so that you can specify rhobar in addition to rho and vfrac.

Right---that's probably the way to go, as we need to output `vfrac`. But if `rhobar` is available then that implies different initialization.

Reply:
Have we created a const only indexer, one that entirely disallows lvalues? We can add that they could potentially be of different types, but I'm not it buys much.

Reply:
Here's what I mean
```C++
template<typename RealIndexer, typename ConstRealIndexer>
void Foo(RealIndexer &&outs, ConstRealIndexer &&ins, int num) {
  portableFor(num, PORTABLE_LAMBDA(int i) { outs[i] = foo(ins[i]); });
}
```
what this buys us is that the input and output arrays can be different types. i.e.,
```C++
Real *my_outs = something;
const Real *my_ins = somethingelse;
Foo(my_outs, my_ins, N);
```
will compile. On the other hand, if `Foo` only templated on `RealIndexer` but didn't allow for a `ConstRealIndexer`, the discarding of the `const` qualifier would mean type inferrence would fail.

Reply:
oh duh, silly me :)

Reply:
a different template parameter has been added in case one passes in const real *'s